File: backend/services/aquarium-service/src/application/services/aquarium.service.test.ts
import { Test, TestingModule } from '@nestjs/testing';
import { NotFoundException, ForbiddenException } from '@nestjs/common';
import { AquariumService } from './aquarium.service';
import { AquariumDomainService } from '../../domain/services/aquarium.domain.service';
import { AquariumRepository } from '../../infrastructure/database/repositories/aquarium.repository';
import { WaterParametersRepository } from '../../infrastructure/database/repositories/water-parameters.repository';
import { EventService } from '../../infrastructure/events/event.service';
import { StorageService } from '../../infrastructure/storage/storage.service';

describe('AquariumService', () => {
  let service: AquariumService;
  let domainService: AquariumDomainService;
  let aquariumRepository: AquariumRepository;
  let waterParametersRepository: WaterParametersRepository;
  let eventService: EventService;
  let storageService: StorageService;

  const mockUserId = 'user123';
  const mockAquariumId = 'aquarium123';

  const mockAquarium = {
    id: mockAquariumId,
    userId: mockUserId,
    name: 'Test Aquarium',
    type: 'freshwater',
    volume: 100,
    volumeUnit: 'liters',
    dimensions: { length: 100, width: 40, height: 50 },
    inhabitants: [],
    equipment: [],
    parameters: {
      temperature: 25,
      ph: 7.0,
      ammonia: 0,
      nitrite: 0,
      nitrate: 10,
    },
    createdAt: new Date(),
    updatedAt: new Date(),
  };

  const mockRepositories = {
    findById: jest.fn(),
    findByUserId: jest.fn(),
    create: jest.fn(),
    update: jest.fn(),
    delete: jest.fn(),
  };

  const mockWaterParametersRepo = {
    create: jest.fn(),
    findByAquariumId: jest.fn(),
    getLatestByAquariumId: jest.fn(),
  };

  const mockEventService = {
    publishEvent: jest.fn(),
  };

  const mockStorageService = {
    uploadImage: jest.fn(),
    deleteImage: jest.fn(),
  };

  const mockDomainService = {
    calculateWaterVolume: jest.fn(),
    validateWaterParameters: jest.fn(),
    checkCompatibility: jest.fn(),
  };

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        AquariumService,
        {
          provide: AquariumDomainService,
          useValue: mockDomainService,
        },
        {
          provide: AquariumRepository,
          useValue: mockRepositories,
        },
        {
          provide: WaterParametersRepository,
          useValue: mockWaterParametersRepo,
        },
        {
          provide: EventService,
          useValue: mockEventService,
        },
        {
          provide: StorageService,
          useValue: mockStorageService,
        },
      ],
    }).compile();

    service = module.get<AquariumService>(AquariumService);
    domainService = module.get<AquariumDomainService>(AquariumDomainService);
    aquariumRepository = module.get<AquariumRepository>(AquariumRepository);
    waterParametersRepository = module.get<WaterParametersRepository>(WaterParametersRepository);
    eventService = module.get<EventService>(EventService);
    storageService = module.get<StorageService>(StorageService);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('create', () => {
    it('should create a new aquarium successfully', async () => {
      const createDto = {
        name: 'New Aquarium',
        type: 'freshwater' as const,
        volume: 100,
        volumeUnit: 'liters' as const,
        dimensions: { length: 100, width: 40, height: 50 },
        description: 'Test aquarium',
      };

      mockDomainService.calculateWaterVolume.mockReturnValue(200);
      mockRepositories.create.mockResolvedValue(mockAquarium);

      const result = await service.create(mockUserId, createDto);

      expect(result).toEqual(mockAquarium);
      expect(mockRepositories.create).toHaveBeenCalledWith({
        ...createDto,
        userId: mockUserId,
        inhabitants: [],
        equipment: [],
        parameters: {},
      });
      expect(mockEventService.publishEvent).toHaveBeenCalledWith('aquarium.created', {
        aquariumId: mockAquarium.id,
        userId: mockUserId,
        name: mockAquarium.name,
      });
    });

    it('should upload image if provided', async () => {
      const createDto = {
        name: 'New Aquarium',
        type: 'freshwater' as const,
        volume: 100,
        volumeUnit: 'liters' as const,
        image: 'base64-image-data',
      };

      const imageUrl = 'https://storage.example.com/image.jpg';
      mockStorageService.uploadImage.mockResolvedValue(imageUrl);
      mockRepositories.create.mockResolvedValue({
        ...mockAquarium,
        imageUrl,
      });

      const result = await service.create(mockUserId, createDto);

      expect(mockStorageService.uploadImage).toHaveBeenCalledWith(
        createDto.image,
        'aquariums',
      );
      expect(result.imageUrl).toBe(imageUrl);
    });
  });

  describe('findAll', () => {
    it('should return all aquariums for a user', async () => {
      const mockAquariums = [mockAquarium, { ...mockAquarium, id: '456' }];
      mockRepositories.findByUserId.mockResolvedValue(mockAquariums);

      const result = await service.findAll(mockUserId);

      expect(result).toEqual(mockAquariums);
      expect(mockRepositories.findByUserId).toHaveBeenCalledWith(mockUserId);
    });

    it('should return empty array if no aquariums found', async () => {
      mockRepositories.findByUserId.mockResolvedValue([]);

      const result = await service.findAll(mockUserId);

      expect(result).toEqual([]);
    });
  });

  describe('findOne', () => {
    it('should return aquarium details with latest parameters', async () => {
      const latestParameters = {
        temperature: 26,
        ph: 7.2,
        ammonia: 0,
        nitrite: 0,
        nitrate: 15,
        recordedAt: new Date(),
      };

      mockRepositories.findById.mockResolvedValue(mockAquarium);
      mockWaterParametersRepo.getLatestByAquariumId.mockResolvedValue(latestParameters);

      const result = await service.findOne(mockUserId, mockAquariumId);

      expect(result).toEqual({
        ...mockAquarium,
        latestParameters,
      });
    });

    it('should throw NotFoundException if aquarium not found', async () => {
      mockRepositories.findById.mockResolvedValue(null);

      await expect(
        service.findOne(mockUserId, mockAquariumId),
      ).rejects.toThrow(NotFoundException);
    });

    it('should throw ForbiddenException if user does not own aquarium', async () => {
      mockRepositories.findById.mockResolvedValue({
        ...mockAquarium,
        userId: 'other-user',
      });

      await expect(
        service.findOne(mockUserId, mockAquariumId),
      ).rejects.toThrow(ForbiddenException);
    });
  });

  describe('recordParameters', () => {
    it('should record water parameters successfully', async () => {
      const parametersDto = {
        temperature: 26,
        ph: 7.2,
        ammonia: 0,
        nitrite: 0,
        nitrate: 15,
      };

      mockRepositories.findById.mockResolvedValue(mockAquarium);
      mockDomainService.validateWaterParameters.mockReturnValue(true);
      mockWaterParametersRepo.create.mockResolvedValue({
        ...parametersDto,
        aquariumId: mockAquariumId,
        recordedAt: new Date(),
      });

      await service.recordParameters(mockUserId, mockAquariumId, parametersDto);

      expect(mockWaterParametersRepo.create).toHaveBeenCalledWith({
        aquariumId: mockAquariumId,
        ...parametersDto,
      });
      expect(mockEventService.publishEvent).toHaveBeenCalledWith(
        'aquarium.parameters_recorded',
        {
          aquariumId: mockAquariumId,
          parameters: parametersDto,
        },
      );
    });

    it('should update aquarium with latest parameters', async () => {
      const parametersDto = {
        temperature: 26,
        ph: 7.2,
      };

      mockRepositories.findById.mockResolvedValue(mockAquarium);
      mockDomainService.validateWaterParameters.mockReturnValue(true);

      await service.recordParameters(mockUserId, mockAquariumId, parametersDto);

      expect(mockRepositories.update).toHaveBeenCalledWith(
        mockAquariumId,
        { parameters: parametersDto },
      );
    });
  });

  describe('addInhabitant', () => {
    it('should add inhabitant successfully', async () => {
      const inhabitantDto = {
        species: 'Neon Tetra',
        quantity: 10,
        type: 'fish' as const,
        addedDate: new Date().toISOString(),
      };

      mockRepositories.findById.mockResolvedValue(mockAquarium);
      mockDomainService.checkCompatibility.mockResolvedValue(true);

      const result = await service.addInhabitant(
        mockUserId,
        mockAquariumId,
        inhabitantDto,
      );

      expect(result.inhabitants).toHaveLength(1);
      expect(result.inhabitants[0]).toMatchObject({
        species: inhabitantDto.species,
        quantity: inhabitantDto.quantity,
        type: inhabitantDto.type,
      });
      expect(mockEventService.publishEvent).toHaveBeenCalledWith(
        'aquarium.inhabitant_added',
        expect.any(Object),
      );
    });

    it('should check compatibility before adding inhabitant', async () => {
      const inhabitantDto = {
        species: 'Goldfish',
        quantity: 1,
        type: 'fish' as const,
      };

      const aquariumWithInhabitants = {
        ...mockAquarium,
        inhabitants: [
          {
            id: '1',
            species: 'Neon Tetra',
            quantity: 10,
            type: 'fish',
          },
        ],
      };

      mockRepositories.findById.mockResolvedValue(aquariumWithInhabitants);
      mockDomainService.checkCompatibility.mockResolvedValue(false);

      await expect(
        service.addInhabitant(mockUserId, mockAquariumId, inhabitantDto),
      ).rejects.toThrow('Inhabitant not compatible with existing inhabitants');
    });
  });

  describe('addEquipment', () => {
    it('should add equipment successfully', async () => {
      const equipmentDto = {
        name: 'Filter',
        type: 'filtration' as const,
        brand: 'Fluval',
        model: 'FX6',
        installedDate: new Date().toISOString(),
      };

      mockRepositories.findById.mockResolvedValue(mockAquarium);

      const result = await service.addEquipment(
        mockUserId,
        mockAquariumId,
        equipmentDto,
      );

      expect(result.equipment).toHaveLength(1);
      expect(result.equipment[0]).toMatchObject({
        name: equipmentDto.name,
        type: equipmentDto.type,
        brand: equipmentDto.brand,
        model: equipmentDto.model,
      });
      expect(mockEventService.publishEvent).toHaveBeenCalledWith(
        'aquarium.equipment_added',
        expect.any(Object),
      );
    });
  });

  describe('delete', () => {
    it('should delete aquarium successfully', async () => {
      mockRepositories.findById.mockResolvedValue(mockAquarium);

      await service.delete(mockUserId, mockAquariumId);

      expect(mockRepositories.delete).toHaveBeenCalledWith(mockAquariumId);
      expect(mockEventService.publishEvent).toHaveBeenCalledWith(
        'aquarium.deleted',
        {
          aquariumId: mockAquariumId,
          userId: mockUserId,
        },
      );
    });

    it('should delete associated image if exists', async () => {
      const aquariumWithImage = {
        ...mockAquarium,
        imageUrl: 'https://storage.example.com/image.jpg',
      };

      mockRepositories.findById.mockResolvedValue(aquariumWithImage);

      await service.delete(mockUserId, mockAquariumId);

      expect(mockStorageService.deleteImage).toHaveBeenCalledWith(
        aquariumWithImage.imageUrl,
      );
    });
  });

  describe('getParametersHistory', () => {
    it('should return parameters history', async () => {
      const mockHistory = [
        {
          temperature: 25,
          ph: 7.0,
          recordedAt: new Date('2024-01-01'),
        },
        {
          temperature: 26,
          ph: 7.2,
          recordedAt: new Date('2024-01-02'),
        },
      ];

      mockRepositories.findById.mockResolvedValue(mockAquarium);
      mockWaterParametersRepo.findByAquariumId.mockResolvedValue(mockHistory);

      const result = await service.getParametersHistory(
        mockUserId,
        mockAquariumId,
        { days: 7 },
      );

      expect(result).toEqual(mockHistory);
      expect(mockWaterParametersRepo.findByAquariumId).toHaveBeenCalledWith(
        mockAquariumId,
        expect.any(Date),
        expect.any(Date),
      );
    });
  });
});

File: mobile/test/features/aquarium/aquarium_service_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/mockito.dart';
import 'package:mockito/annotations.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';

import '../../../lib/features/aquarium/services/aquarium_service.dart';
import '../../../lib/features/aquarium/models/aquarium_model.dart';
import '../../../lib/core/api/api_client.dart';
import '../../../lib/core/storage/storage_service.dart';

@GenerateMocks([ApiClient, StorageService, http.MultipartRequest])
import 'aquarium_service_test.mocks.dart';

void main() {
  late AquariumService aquariumService;
  late MockApiClient mockApiClient;
  late MockStorageService mockStorageService;

  setUp(() {
    mockApiClient = MockApiClient();
    mockStorageService = MockStorageService();
    aquariumService = AquariumService();
    // Inject mocks
    aquariumService.apiClient = mockApiClient;
    aquariumService.storageService = mockStorageService;
  });

  group('AquariumService', () {
    group('getAquariums', () {
      test('should return list of aquariums on success', () async {
        // Arrange
        final mockResponse = [
          {
            'id': '1',
            'name': 'Test Aquarium',
            'type': 'freshwater',
            'volume': 100.0,
            'volumeUnit': 'liters',
            'createdAt': DateTime.now().toIso8601String(),
            'updatedAt': DateTime.now().toIso8601String(),
          }
        ];

        when(mockApiClient.get('/aquariums')).thenAnswer(
          (_) async => {'data': mockResponse},
        );

        // Act
        final result = await aquariumService.getAquariums();

        // Assert
        expect(result, hasLength(1));
        expect(result[0].name, 'Test Aquarium');
        expect(result[0].type, WaterType.freshwater);
        verify(mockApiClient.get('/aquariums')).called(1);
      });

      test('should throw exception on API error', () async {
        // Arrange
        when(mockApiClient.get('/aquariums')).thenThrow(
          Exception('Network error'),
        );

        // Act & Assert
        expect(
          () => aquariumService.getAquariums(),
          throwsException,
        );
      });
    });

    group('createAquarium', () {
      test('should create aquarium successfully', () async {
        // Arrange
        final newAquarium = Aquarium(
          id: '',
          name: 'New Tank',
          type: WaterType.freshwater,
          volume: 200,
          volumeUnit: VolumeUnit.liters,
          createdAt: DateTime.now(),
          updatedAt: DateTime.now(),
        );

        final mockResponse = {
          'id': '123',
          'name': 'New Tank',
          'type': 'freshwater',
          'volume': 200.0,
          'volumeUnit': 'liters',
          'createdAt': DateTime.now().toIso8601String(),
          'updatedAt': DateTime.now().toIso8601String(),
        };

        when(mockApiClient.post('/aquariums', body: anyNamed('body')))
            .thenAnswer((_) async => {'data': mockResponse});

        // Act
        final result = await aquariumService.createAquarium(newAquarium);

        // Assert
        expect(result.id, '123');
        expect(result.name, 'New Tank');
        verify(mockApiClient.post('/aquariums', body: anyNamed('body')))
            .called(1);
      });

      test('should handle image upload when creating aquarium', () async {
        // Arrange
        final newAquarium = Aquarium(
          id: '',
          name: 'New Tank',
          type: WaterType.freshwater,
          volume: 200,
          volumeUnit: VolumeUnit.liters,
          createdAt: DateTime.now(),
          updatedAt: DateTime.now(),
        );

        final mockImagePath = '/path/to/image.jpg';
        final mockImageUrl = 'https://storage.example.com/image.jpg';

        when(mockApiClient.uploadFile(
          any,
          any,
          field: anyNamed('field'),
          method: anyNamed('method'),
          data: anyNamed('data'),
        )).thenAnswer((_) async => {'imageUrl': mockImageUrl});

        // Act
        await aquariumService.createAquarium(newAquarium, imagePath: mockImagePath);

        // Assert
        verify(mockApiClient.uploadFile(
          any,
          any,
          field: 'image',
          method: 'POST',
          data: anyNamed('data'),
        )).called(1);
      });
    });

    group('updateAquarium', () {
      test('should update aquarium successfully', () async {
        // Arrange
        final aquarium = Aquarium(
          id: '123',
          name: 'Updated Tank',
          type: WaterType.marine,
          volume: 300,
          volumeUnit: VolumeUnit.gallons,
          createdAt: DateTime.now(),
          updatedAt: DateTime.now(),
        );

        when(mockApiClient.put('/aquariums/123', body: anyNamed('body')))
            .thenAnswer((_) async => {'data': aquarium.toJson()});

        // Act
        final result = await aquariumService.updateAquarium(aquarium);

        // Assert
        expect(result.name, 'Updated Tank');
        expect(result.type, WaterType.marine);
        verify(mockApiClient.put('/aquariums/123', body: anyNamed('body')))
            .called(1);
      });
    });

    group('deleteAquarium', () {
      test('should delete aquarium successfully', () async {
        // Arrange
        when(mockApiClient.delete('/aquariums/123'))
            .thenAnswer((_) async => {});

        // Act
        await aquariumService.deleteAquarium('123');

        // Assert
        verify(mockApiClient.delete('/aquariums/123')).called(1);
      });
    });

    group('recordParameters', () {
      test('should record water parameters successfully', () async {
        // Arrange
        final parameters = WaterParameters(
          temperature: 25.5,
          ph: 7.2,
          ammonia: 0.0,
          nitrite: 0.0,
          nitrate: 10.0,
          recordedAt: DateTime.now(),
        );

        when(mockApiClient.post(
          '/aquariums/123/parameters',
          body: anyNamed('body'),
        )).thenAnswer((_) async => {'data': parameters.toJson()});

        // Act
        await aquariumService.recordParameters('123', parameters);

        // Assert
        verify(mockApiClient.post(
          '/aquariums/123/parameters',
          body: anyNamed('body'),
        )).called(1);
      });

      test('should cache parameters locally', () async {
        // Arrange
        final parameters = WaterParameters(
          temperature: 25.5,
          ph: 7.2,
          recordedAt: DateTime.now(),
        );

        when(mockApiClient.post(any, body: anyNamed('body')))
            .thenAnswer((_) async => {'data': parameters.toJson()});
        when(mockStorageService.setString(any, any))
            .thenAnswer((_) async => true);

        // Act
        await aquariumService.recordParameters('123', parameters);

        // Assert
        verify(mockStorageService.setString(
          'aquarium_123_latest_parameters',
          any,
        )).called(1);
      });
    });

    group('getParameterHistory', () {
      test('should fetch parameter history', () async {
        // Arrange
        final mockHistory = [
          {
            'temperature': 25.0,
            'ph': 7.0,
            'recordedAt': DateTime.now().subtract(Duration(days: 1)).toIso8601String(),
          },
          {
            'temperature': 26.0,
            'ph': 7.2,
            'recordedAt': DateTime.now().toIso8601String(),
          },
        ];

        when(mockApiClient.get('/aquariums/123/parameters/history'))
            .thenAnswer((_) async => {'data': mockHistory});

        // Act
        final result = await aquariumService.getParameterHistory('123');

        // Assert
        expect(result, hasLength(2));
        expect(result[0].temperature, 25.0);
        expect(result[1].temperature, 26.0);
      });
    });

    group('addInhabitant', () {
      test('should add inhabitant successfully', () async {
        // Arrange
        final inhabitant = Inhabitant(
          id: '',
          species: 'Neon Tetra',
          quantity: 10,
          type: InhabitantType.fish,
          addedDate: DateTime.now(),
        );

        final mockResponse = {
          'id': 'inhabitant123',
          'species': 'Neon Tetra',
          'quantity': 10,
          'type': 'fish',
          'addedDate': DateTime.now().toIso8601String(),
        };

        when(mockApiClient.post(
          '/aquariums/123/inhabitants',
          body: anyNamed('body'),
        )).thenAnswer((_) async => {'data': mockResponse});

        // Act
        final result = await aquariumService.addInhabitant('123', inhabitant);

        // Assert
        expect(result.id, 'inhabitant123');
        expect(result.species, 'Neon Tetra');
      });
    });

    group('addEquipment', () {
      test('should add equipment successfully', () async {
        // Arrange
        final equipment = Equipment(
          id: '',
          name: 'Fluval FX6',
          type: EquipmentType.filter,
          brand: 'Fluval',
          model: 'FX6',
          installedDate: DateTime.now(),
        );

        final mockResponse = {
          'id': 'equipment123',
          'name': 'Fluval FX6',
          'type': 'filter',
          'brand': 'Fluval',
          'model': 'FX6',
          'installedDate': DateTime.now().toIso8601String(),
        };

        when(mockApiClient.post(
          '/aquariums/123/equipment',
          body: anyNamed('body'),
        )).thenAnswer((_) async => {'data': mockResponse});

        // Act
        final result = await aquariumService.addEquipment('123', equipment);

        // Assert
        expect(result.id, 'equipment123');
        expect(result.name, 'Fluval FX6');
      });
    });

    group('getAquariumHealth', () {
      test('should calculate aquarium health score', () async {
        // Arrange
        final aquarium = Aquarium(
          id: '123',
          name: 'Test Tank',
          type: WaterType.freshwater,
          volume: 100,
          volumeUnit: VolumeUnit.liters,
          parameters: WaterParameters(
            temperature: 25.0,
            ph: 7.0,
            ammonia: 0.0,
            nitrite: 0.0,
            nitrate: 10.0,
            recordedAt: DateTime.now(),
          ),
          createdAt: DateTime.now(),
          updatedAt: DateTime.now(),
        );

        when(mockApiClient.get('/aquariums/123'))
            .thenAnswer((_) async => {'data': aquarium.toJson()});

        // Act
        final health = await aquariumService.getAquariumHealth('123');

        // Assert
        expect(health.score, greaterThanOrEqualTo(0));
        expect(health.score, lessThanOrEqualTo(100));
        expect(health.status, isNotEmpty);
      });
    });
  });
}